keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+  at all that a driver is using semantic parameters under the hood.
+
+### Non-Goals
+
+- Scheduling performance is expected to become better compared to using the
+  PodSchedulingContext. However, this is not the reason for this KEP.
+
+
+## Proposal
+
+### Publishing node resources
+
+The kubelet publishes NodeResourceSlices to the API server with content
+provided by DRA drivers running on each node. Access control through the node
+authorizer ensures that the kubelet running on one node is not allowed to
+create or modify NodeResourceSlices belonging to other nodes.
+
+NodeResourceSlices are published separately for each driver, using a version of
+the API that matches what that driver is using. This implies that the kubelet, DRA
+driver, and apiserver must all support the *same* API version. It might be
+possible to support version skew (= keeping kubelet at an older version than
+the control plane and the DRA drivers) in the future, but currently this is out
+of scope.
+
+Embedded inside a NodeResourceSlice is the representation of the resources
+managed by a driver. With the "semantic model" of `PartitionableResources` this
+takes the form:
+
+```yaml
+kind: NodeResourceSlice
+apiVersion: resource.k8s.io/v1alpha2
+...
+spec:
+  nodeName: worker-1
+  driverName: cards.dra.example.com
+  partitionable:
+    ...
+```
+
+Here all resources associated with the `cards.dra.example.com` driver are to
+be placed under the `partitionable` field in the `NodeResourceSlice` object. As
+we add more "semantic models" in the future, *alternate* fields will be added
+at the same level as `partitionable`, and driver implementors will be able to
+choose which one they want to use to represent their resources.
+
+Below is an example of a driver that provides two discrete GPU cards using the
+`partitionable` model. The first card can only be used as a whole. The second
+card can be used as a whole, split up into four equal pieces, in halves, or one
+half and two quarters:
+
+```yaml
+kind: NodeResourceSlice
+apiVersion: resource.k8s.io/v1alpha2
+...
+spec:
+  nodeName: worker-1
+  driverName: cards.dra.example.com
+  partitionable:
+    commonAttributes:
+    - name: drivers
+      attributes:
+      - name: driverVersion
+        String: 1.2.3
+      - name: runtimeVersion
+        String: 11.1.42
+    - name: t1000-gpu
+      attributes:
+      - name: type
+        string: GPU
+      - name: memory
+        quantity: 16Gi
+      - name: productName:
+        string: ACME T1000 32GB
+    - name: a4-gpu
+      attributes:
+      - name: type
+        String: GPU
+      - name: memory
+        quantity: 32Gi
+      - name: productName:
+        String: ACME A4-PCIE-40GB
+    - name: a4-half
+      attributes:
+      - name: memory
+        quantity: 16Gi
+    - name: a4-quarter
+      attributes:
+      - name: memory
+        quantity: 8Gi
+
+    resourceInstances:
+    - name: gpu-0
+      commonAttributesRefs:
+      - drivers
```

Comment:
We can argue about the structure of this during the implementation, but the reason I thought mapping a resources to a named list of attributes was nice, was because then the scheduler doesn't actually *need* to know anything about the structure of these resources. All of the translations from vendor-specific CRDs to the generic `ResourceClaimParameters` object become nothing more than a set of boolean expressions on a set of attributes. If an instance's attributes match that expression -- it is a candidate for allocation, if they don't -- it doesn't.

Reply:
English is hard.  I whipped this up:

https://go.dev/play/p/RJ8uGgBNSvn

I hyper-normalized it, so almost nothing get repeated.  I think it might be more readable it "shape" was just part of the definition of "mode".

Reply:
I see what you've done -- basically try and enforce some structure around where the attributes are applied, rather than have them be loosely defined as untyped "groups" of attributes.

However, I don't quite see the advantage of doing this, especially in light of this comment you have:
> // Scheduler can flatten all the instances' attributes, run the Constraints expression, then pick a best-fit from the candidates.

If this structure is not for the scheduler, then who is it for? The driver implementor?

I wouldn't necessarily be opposed to adding structure such as this if it means we can reach consensus, but I'd almost rather just do away with *any* notion of named attribute "groups" (or shapes or classes) and instead just have a flat list of attributes attached to each instance (even if it means those attributes are repeated), i.e.:
```
type ResourceInstance struct {
	Name       string
	Attributes []ResourceAttribute
	Modes      []ResourceMode
}

type ResourceMode struct {
	Name      string
	Instances []ResourceInstance
}

type ResourceAttribute struct {
	Name string
	// One of the following:
	String   *string   `json:",omitempty"`
	Int      *int      `json:",omitempty"`
	Quantity *Quantity `json:",omitempty"`
}
```

Reply:
This matches the intended semantic. I don't have a strong opinion whether we should literally express it like that in the NodeResourceSlice.

This is a slight tangent, but if we can, I would write it as literal map:
```
Attributes map[string]ResourceAttribute

type ResourceAttribute struct {
	// One of the following:
	String   *string   `json:",omitempty"`
	Int      *int      `json:",omitempty"`
	Quantity *Quantity `json:",omitempty"`
}
```

Maps are discouraged in the Kubernetes API in favor of lists, but we don't need the ability to add or remove entries and maps would be simpler for the implementation.

Reply:
> If this structure is not for the scheduler, then who is it for?

It's for the sake of efficiency.  We could fully denormalize it and it would be easier to use, but very repetitive.  I was trying to balance DRY with usability.  Each "instance" has a clear lineage of attributes, which map to how a driver author would probably structure things and are easy to flatten.  That said, a driver is almost certainly going to use `const`s so the risk of mistakes is low.  This could end up a fairly large object if there are many attributes * many modes * many shapes * many instances